### PR TITLE
cuDNN Deterministic mode

### LINKED ISSUE #2574: Add chainer.config.deterministic
Related to #2048

### PR DESCRIPTION
**Pull request by delta2323**
This PR adds `configuration.config.deterministic`.

Fix #2574, related to #2048

### CODE REVIEW

**Comment by unnonouno** [comment]
Check test result.

**Comment by beam2d** [comment]
Do you think it's better to raise an error in the deterministic mode if any random functions are called (e.g. dropout)?

**Comment by delta2323** [comment]
I have no clear answer for now. But, I have two comments:
* In order to answer the question, we first need to clarify the use case, or why users want to use the deterministic mode.
* At the extreme, if we forbid any random behavior in the deterministic mode, we cannot use functions that is random by nature (e.g. sampling function like `F.gaussian`).

**Comment by niboshi** [comment]
Could you resolve conflicts and fix test failures?

As for the discussion above, if I were a user and saw a global configuration named `deterministic`, I would expect all the behavior within chainer is deterministic, i.e. the net result is the same every time.

**Comment by niboshi** [comment]
One possible solution I can think of is to make it more specific, i.e. rename `deterministic` to `deterministic_cudnn`, for example.

**Comment by delta2323** [comment]
Definitely it is beneficial for users if the framework ensures all Chainer functions behave deterministically (or raise appropriate errors) in the deterministic mode. But I am concerning the implementation cost (mainly writing tests) to ensure it.

On the other hand, I cannot evaluate to what extent `deterministic_cudnn` is useful for users.

**Comment by niboshi** [comment]
Maybe `deterministic_cudnn` would be meaningless if there were *true* `deterministic` mode.
But if we did not offer such mode, users could use `deterministic_cudnn` mode if they knew cuDNN is the only probabilistic part in their networks.

IMO the reasonable choice would be either
* Offer *true* `deterministic` mode which raises error (or at least warning) on every probabilistic operations, or
* Offer `deterministic_cudnn` mode and let users responsible for other probabilistic operations.

**Comment by beam2d** [comment]
Note that the non-deterministic behavior appears whenever we use atomic operations in their kernels. AFAIK, `scatter_add` uses atomic operations, and `__getitem__` function calls it in the backward. I don't know if we will provide a deterministic implementation for it, but it may happen (I mean, this is a case outside of cuDNN).

**Comment by delta2323** [comment]
I am wondering if we can identify the all source of stochasticity. I am not familiar with this kind of problem but my intuition is it is extremely difficult. But cuDNN does it because it provides deterministic option. So, is there a way to achieve it somehow?

**Comment by niboshi** [comment]
FYI, TensorFlow gave up determinism on the same (I think) problem as above, at least for now: https://github.com/tensorflow/tensorflow/issues/2652. I think it's almost impossible to offer *true* determinism with limited resource.

Another FYI, Theano has very specific flags to control cuDNN determinism: `THEANO_FLAGS="dnn.conv.algo_bwd_filter=deterministic,dnn.conv.algo_bwd_data=deterministic" `.

I'm beginning to think the original `deterministic` mode (with incomplete determinism) is also acceptable *if* what will become deterministic is carefully documented.

**Comment by delta2323** [comment]
@niboshi Thank you for your information. Considering the difficulty of supporting full deterministic mode and current goal, I think it is a good compromise to offer the determinism limited to cuDNN with more specific name (`cudnn_deterministic`).

@beam2d By the way, do you have a guideline whether we should provide some mode via a global config variable or an argument of functions?

**Comment by beam2d** [comment]
I don't have a concrete guideline, but I can say that the mode should be provided as a global flag if one who writes the model and one who wants to configure the flag are different people. In this case, the global config can be changed without editing the model definition, whereas in the case of using an argument the flag needs to be changed by editing the model definition (or the model author has to provide a dedicated way to change the mode from the outside of the model definition, e.g. by exposing the flag as an attribute of the link). Another point: if the flag is usually modified differently for every function call, it is better to be provided as an argument. In this case, the argument approach makes the code simpler (if we use a global config, the model author has to wrap every function call by `with using_config(...)`, which messes up the code).

**Comment by delta2323** [comment]
@beam2d Thank you for answering my (somewhat abstract) questions. Your sufficient conditions are reasonable to me. For example, `train` and `(cudnn_)deterministic` modes satisfy the first condition, if my understanding is correct.

File: tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py [hunk 1]
```diff
@@ -177,21 +186,33 @@ def forward(self):
...
-            y.grad = self.gy
-            if cuda.cudnn.cudnn.getVersion() >= 3000:
-                name = 'cupy.cudnn.cudnn.convolutionBackwardData_v3'
-            else:
-                name = 'cupy.cudnn.cudnn.convolutionBackwardData_v2'
-            with mock.patch(name) as func:
-                y.backward()
-                self.assertEqual(func.called, self.expect)
+            with (chainer.using_config(
+                    'cudnn_deterministic', self.cudnn_deterministic)):
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py]: Ditto (outer parentheses)

File: tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py [hunk 2]
```diff
@@ -80,18 +82,22 @@ def test_forward_consistency(self):
         x_cpu = chainer.Variable(self.x)
         W_cpu = chainer.Variable(self.W)
         b_cpu = None if self.nobias else chainer.Variable(self.b)
-        y_cpu = F.deconvolution_2d(
-            x_cpu, W_cpu, b_cpu, stride=self.stride, pad=self.pad,
-            outsize=self.outsize)
+        with (chainer.using_config(
+                'cudnn_deterministic', self.cudnn_deterministic)):
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py]: Ditto (outer parenthese)

File: tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py [hunk 3]
```diff
@@ -80,18 +82,22 @@ def test_forward_consistency(self):
...
         x_gpu = chainer.Variable(cuda.to_gpu(self.x))
         W_gpu = chainer.Variable(cuda.to_gpu(self.W))
         b_gpu = None if self.nobias else chainer.Variable(
             cuda.to_gpu(self.b))
         with chainer.using_config('use_cudnn', self.use_cudnn):
-            y_gpu = F.deconvolution_2d(
-                x_gpu, W_gpu, b_gpu, stride=self.stride, pad=self.pad,
-                outsize=self.outsize)
+            with (chainer.using_config(
+                    'cudnn_deterministic', self.cudnn_deterministic)):
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py]: Ditto (outer parentheses)

File: tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py [hunk 4]
```diff
@@ -123,10 +129,12 @@ def check_backward(self, x_data, W_data, b_data, y_grad):
             args = args + (b_data,)
 
         with chainer.using_config('use_cudnn', self.use_cudnn):
-            gradient_check.check_backward(
-                deconvolution_2d.Deconvolution2DFunction(
-                    self.stride, self.pad, self.outsize),
-                args, y_grad, **self.check_backward_options)
+            with (chainer.using_config(
+                    'cudnn_deterministic', self.cudnn_deterministic)):
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py]: Ditto (outer parentheses)

File: tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py [hunk 5]
```diff
@@ -181,26 +190,38 @@ def test_call_cudnn_forward(self):
         else:
             name = 'cupy.cudnn.cudnn.convolutionBackwardData_v2'
         with chainer.using_config('use_cudnn', self.use_cudnn):
-            with mock.patch(name) as func:
-                self.forward()
-                self.assertEqual(func.called, self.expect)
+            with (chainer.using_config(
+                    'cudnn_deterministic', self.cudnn_deterministic)):
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py]: Ditto (outer parentheses)

File: tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py [hunk 6]
```diff
@@ -181,26 +190,38 @@ def test_call_cudnn_forward(self):
...
 
     def test_call_cudnn_backward(self):
         with chainer.using_config('use_cudnn', self.use_cudnn):
-            y = self.forward()
-            y.grad = self.gy
-            with mock.patch('cupy.cudnn.cudnn.convolutionForward') as func:
-                y.backward()
-                self.assertEqual(func.called, self.expect)
+            with (chainer.using_config(
+                    'cudnn_deterministic', self.cudnn_deterministic)):
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py]: Ditto (outer parentheses)

File: tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py [hunk 7]
```diff
@@ -59,17 +61,21 @@ def test_forward_consistency(self, nobias=False):
         x_cpu = chainer.Variable(self.x)
         W_cpu = chainer.Variable(self.W)
         b_cpu = None if nobias else chainer.Variable(self.b)
-        y_cpu = functions.convolution_2d(
-            x_cpu, W_cpu, b_cpu, stride=self.stride, pad=self.pad,
-            cover_all=self.cover_all)
+        with (chainer.using_config(
+                'cudnn_deterministic', self.cudnn_deterministic)):
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py]: I don't think you need outer paretheses here.

File: tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py [hunk 8]
```diff
@@ -59,17 +61,21 @@ def test_forward_consistency(self, nobias=False):
...
 
         x_gpu = chainer.Variable(cuda.to_gpu(self.x))
         W_gpu = chainer.Variable(cuda.to_gpu(self.W))
         b_gpu = None if nobias else chainer.Variable(cuda.to_gpu(self.b))
         with chainer.using_config('use_cudnn', self.use_cudnn):
-            y_gpu = functions.convolution_2d(
-                x_gpu, W_gpu, b_gpu, stride=self.stride, pad=self.pad,
-                cover_all=self.cover_all)
+            with (chainer.using_config(
+                    'cudnn_deterministic', self.cudnn_deterministic)):
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py]: Ditto (outer parentheses)

File: tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py [hunk 9]
```diff
@@ -104,10 +110,12 @@ def check_backward(self, x_data, W_data, b_data, y_grad):
             args = args + (b_data,)
 
         with chainer.using_config('use_cudnn', self.use_cudnn):
-            gradient_check.check_backward(
-                convolution_2d.Convolution2DFunction(
-                    self.stride, self.pad, self.cover_all),
-                args, y_grad, **self.check_backward_options)
+            with (chainer.using_config(
+                    'cudnn_deterministic', self.cudnn_deterministic)):
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py]: Ditto (outer parentheses)

File: chainer/functions/connection/convolution_2d.py [hunk 10]
```diff
@@ -229,37 +237,43 @@ def backward_gpu(self, inputs, grad_outputs):
...
                     algo, workspace.data.ptr, workspace_size,
                     zero.data, x_desc.value, gx.data.ptr)
             else:
-                if self.deterministic:
-                    raise ValueError("'deterministic' option not available "
-                                     "for cuDNN versions < v3")
+                if configuration.config.cudnn_deterministic:
+                    raise ValueError(
+                        "`cudnn_deterministic` option must be False "
+                        "if the back propagation of "
```
Comment (niboshi) [inline chainer/functions/connection/convolution_2d.py]: back propagation -> backpropagation
or "backward computation"?

File: chainer/functions/connection/deconvolution_2d.py [hunk 11]
```diff
@@ -280,23 +288,29 @@ def backward_gpu(self, inputs, grad_outputs):
...
                     algo, workspace.data.ptr, workspace_size,
                     zero.data, self.filter_desc.value, gW.data.ptr)
             else:
-                if self.deterministic:
-                    raise ValueError("'deterministic' option not available "
-                                     "for cuDNN versions < v3")
+                if configuration.config.cudnn_deterministic:
+                    raise ValueError(
+                        "`cudnn_deterministic` option must be False "
+                        "if the back propagation of "
```
Comment (niboshi) [inline chainer/functions/connection/deconvolution_2d.py]: back propagation -> backpropagation
or "backward computation"?

**Comment by niboshi** [comment]
Can you fix test error?
```
FAIL: test_backward_gpu (test_convolution_2d.TestConvolution2DFunction_param_2)  parameter: 
{'cudnn_deterministic': True, 'c_contiguous': True, 'cover_all': False,
 'W_dtype': <type 'numpy.float32'>, 'x_dtype': <type 'numpy.float32'>}

:

ValueError: `cudnn_deterministic` option must be False if the backpropagation of 
chainer.functions.Convolution2D uses cuDNN and cuDNN versions < v3.
 Turn off cudnn_deterministic option with `chainer.using_config('cudnn_deterministic', False)` context.
```

Perhaps you need `if self.cudnn_version < 3000:` condition in the test?

**Comment by delta2323** [comment]
I want to confirm my understanding. We have to check the value of `cudnn_deterministic` is `False` [here](https://github.com/pfnet/chainer/blob/master/chainer/functions/connection/deconvolution_2d.py#L173), too, right?

**Comment by delta2323** [comment]
@niboshi  GPU test passed. Could you check the PR again?

**Comment by niboshi** [comment]
Thank you. LGTM.

**Pull request merged**